debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerStreamingChangeEventSource.java

```diff
@@ -346,5 +354,54 @@ public String toString() {
             return "ChangeTablePointer [changeTable=" + changeTable + ", resultSet=" + resultSet + ", completed="
                     + completed + ", currentChangePosition=" + currentChangePosition + "]";
         }
+
+        private RowMapper createRowMapper(Table table) throws SQLException {
+            final List<String> sourceTableColumns = table.columnNames();
+            final List<String> resultColumns = getResultColumnNames();
+            final int sourceColumnCount = sourceTableColumns.size();
+            final int resultColumnCount = resultColumns.size();
+
+            if (sourceTableColumns.equals(resultColumns)) {
+                return resultSet -> {
+                    final Object[] data = new Object[sourceColumnCount];
+                    for (int i = 0; i < sourceColumnCount; i++) {
+                        data[i] = resultSet.getObject(COL_DATA + i);
+                    }
+                    return data;
+                };
+            }
+            else {
+                final Map<Integer, Integer> indicesMapping =
+                        getIndicesMapping(sourceTableColumns, resultColumns);
+                return resultSet -> {
+                    final Object[] data = new Object[sourceColumnCount];
+                    for (int i = 0; i < resultColumnCount; i++) {
+                        int index = indicesMapping.get(i);
+                        data[index] = resultSet.getObject(COL_DATA + i);
+                    }
+                    return data;
+                };
+            }
+        }
+
+        private List<String> getResultColumnNames() throws SQLException {
+            final int columnCount = resultSet.getMetaData().getColumnCount() - (COL_DATA - 1);
+            final List<String> columns = new ArrayList<>(columnCount);
+            for (int i = 0; i < columnCount; ++i) {
+                columns.add(resultSet.getMetaData().getColumnName(COL_DATA + i));
+            }
+            return columns;
+        }
+
+        private Map<Integer, Integer> getIndicesMapping(List<String> sourceTableColumns,
```

Comment:
Can you add a dedicated type:

```
class IndicesMapping {
    int getSourceColumnIndex(int resultCaptureInstanceColumnIndex) { ... }
}
```

Makes it all a bit easier to grasp.

Reply:
Done